### PR TITLE
Add authentication to website

### PR DESCRIPTION
**Pull request by Nicholas2750**
- Add utility class that can handle all authentication functionality from now on (login, logout, get LDAP, get email, etc.).
- Add the ability for child classes of DataServlet to be able to attach arbitrary extra parameters that are not included in requests, allowing more flexibility and sophisticated behavior.
- Add functionality in CommentData classes to assert user privilege level and adjust the data it sends subsequently.
- Add UI changes to include login / logout button and warning when not logged in with @google email.
- Add logic in both client and server side such that comment section can only be used by Googlers. LDAP will now be used as username in comment sections.

Screenshots:
- https://screenshot.googleplex.com/uKzpzrdDGDE.png (before logging in)
- https://screenshot.googleplex.com/U81nQEhoYPS.png (logging in with non @google email)
- https://screenshot.googleplex.com/wfOqtrTORwg.png (logging in with @google email)

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/AuthCheck.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.gson.Gson;
+
+/**
```
Comment (chronologos) [inline portfolio/src/main/java/com/google/sps/servlets/AuthCheck.java:12]: i think the code is well-organized. 

Possible improvements to consider:

1. Why is the utility class (which interacts with appengine users api) bundled together with a http handler?

2. Side note for this design: why do we want to allow inheritance on a handler? Is there well-defined behavior in instances where /some-handler is defined by a parent and then inherited?

Comment (Nicholas2750) [reply portfolio/src/main/java/com/google/sps/servlets/AuthCheck.java]: Thanks for the feedback!

1. Great point! I initially wanted to split the http handler and the utility functions into two separate classes. However, I thought that the nature of the code in that class is coupled too tightly that I decided to keep them together.
After more reconsideration though, I realized that making an entire new utils section would make the codebase clearer. So, I moved the utility `AuthCheck` functions to `utils/AuthCheck.java`. I also moved my `GSON` library wrapper to `utils/Parse.java`.

2. There is a well-defined behavior for inheriting handler. The `@WebServlet()` annotation only serves to mark that any GET / POST function to the URL will be handled by the `doGet()` and `doPost()` in that class. For example, if a `Parent` class has `@WebServlet(/parent)`, we can inherit it into `Child` class with say `WebServlet(/child)` and they'll simply be serving two different endpoints. All requests to `/parent` is handled by `Parent` class and the same goes with `/child` and `Child`. However, `Child` now has access to all protected / public methods and variables inside `Parent`. This allows us to re-use all relevant methods and variables. We can also take advantage of Java's method overloading like how the `doGet() in `CommentDataServlet` can re-use the `doGet()` in DataServlet`:
```
@WebServlet("/comment-data")
public class CommentDataServlet extends DataServlet {
  private final int MIN_COMMENTS_NUMBER         = 5;
  private final String COMMENT_NUMBER_PARAMETER = "comments-number";
  private final String USER_LDAP_PARAMETER      = "ldap";
  protected final String ENTITY_KIND            = "Comment";

  @Override
  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
    if (!hasEnoughPrivilege()) {
      return;
    }

    int commentsNumber  = parseIntParameter(request, COMMENT_NUMBER_PARAMETER);
    commentsNumber      = Math.max(commentsNumber, MIN_COMMENTS_NUMBER);

    // This is where we can take advantage of a much more general doGet() declared in DataServlet that we inherited
    doGet(response, ENTITY_KIND, ENTITY_TIMESTAMP_PARAMETER, commentsNumber);
  }
}
```

Comment (chronologos) [reply portfolio/src/main/java/com/google/sps/servlets/AuthCheck.java]: great changes and a good explanation. thank you!

Comment (Nicholas2750) [reply portfolio/src/main/java/com/google/sps/servlets/AuthCheck.java]: Thanks for the review!

**Pull request merged**